kuma/api/v1/plus/bookmarks.py

```diff
@@ -0,0 +1,190 @@
+        # ).exists():
+        #     return HttpResponseForbidden("not a subscriber")
+        return func(request, *args, **kwargs)
+
+    return inner
+
+
+@never_cache
+@require_http_methods(["GET"])
+@is_subscriber
+def bookmarks(request):
+    try:
+        page = int(request.GET.get("page") or "1")
+        assert page > 0 and page < 100
+    except (ValueError, AssertionError):
+        return HttpResponseBadRequest("invalid 'page'")
+    try:
+        per_page = int(
+            request.GET.get("per_page") or settings.API_V1_BOOKMARKS_PAGE_SIZE
+        )
+        assert per_page > 0 and per_page <= 100
+    except (ValueError, AssertionError):
+        return HttpResponseBadRequest("invalid 'per_page'")
+
+    users_bookmarks = Bookmark.objects.filter(
+        user_id=request.user.id, deleted__isnull=True
+    ).order_by("-created")
+
+    def serialize_bookmarks(qs):
+        for bookmark in qs.select_related("documenturl"):
+            # In Yari, the `metadata["parents"]` will always include "self"
+            # in the last link. Omit that here.
+            parents = bookmark.documenturl.metadata.get("parents", [])[:-1]
+            yield {
+                "id": bookmark.id,
+                "url": bookmark.documenturl.metadata["mdn_url"],
+                "title": bookmark.documenturl.metadata["title"],
+                "parents": parents,
+                "created": bookmark.created,
+            }
+
+    m = (page - 1) * per_page
+    n = page * per_page
+    context = {
+        "items": list(serialize_bookmarks(users_bookmarks[m:n])),
+        "metadata": {
+            "total": users_bookmarks.count(),
+            "page": page,
+            "per_page": per_page,
+        },
+    }
+    context["csrfmiddlewaretoken"] = get_token(request)
+    return JsonResponse(context)
+
+
+@never_cache
+@require_http_methods(["GET", "POST"])
+@is_subscriber
+def bookmarked(request):
+
+    url = (
+        (request.GET.get("url") or "")
+        .strip()
+        .replace("https://developer.mozilla.org", "")
+    )
+    if not url:
+        return HttpResponseBadRequest("missing 'url'")
+    if not valid_url(url):
+        return HttpResponseBadRequest("invalid 'url'")
+
+    absolute_url = f"{settings.BOOKMARKS_BASE_URL}{url}/index.json"
+
+    if request.method == "POST":
+        # Let's toggle!
+        try:
+            documenturl = DocumentURL.objects.get(uri=DocumentURL.normalize_uri(url))
+            assert not documenturl.invalid
+        except DocumentURL.DoesNotExist:
+            response = download_url(absolute_url)
+            metadata = response.json()["doc"]
+            # Because it's so big, don't store certain fields that are
+            # not helpful.
+            metadata.pop("body", None)
+            metadata.pop("toc", None)
+            metadata.pop("sidebarHTML", None)
+            metadata.pop("other_translations", None)
+            metadata.pop("flaws", None)
```

Comment:
We're only using the `title`, `mdn_url`, and `parents` from `metadata`. I think we should only store what we're going to use. Also, how do `DocumentURL` objects get updated, for example if the document they represent is moved?

Reply:
I changed it to an "allowlist" of keys instead. 



> Also, how do `DocumentURL` objects get updated, for example if the document they represent is moved?

They don't in this first version. 
I was hoping we add a something on the message queue to improve that. 
For example, we could have a periodic job that loops over the least recently checked `DocumentURL` rows and updates them. 
Another option is that we do something like this:

```diff
if not DocumentURL.objects.filter(url=ur).exists():
    return DocumentURL.objects.create(...)
-return DocumentURL.objects.get(url=url)
+else:
+    doc = DocumentURL.objects.get(url=url)
+    if (timezone.now() - doc.updated_at).days > 24 * 60 * 60:
+        update_documenturl(doc.id)  # send to a Celery task
+    return doc
```

Or, if we want to avoid Celery entirely, we can boldly go ahead and make it sync like this:

```diff
if not DocumentURL.objects.filter(url=ur).exists():
    return DocumentURL.objects.create(...)
-return DocumentURL.objects.get(url=url)
+else:
+    doc = DocumentURL.objects.get(url=url)
+    if (timezone.now() - doc.updated_at).days > 24 * 60 * 60:
+        doc.update()  # sync refresh 
+        doc.refresh_from_db()
+    return doc
```

if we do that, it's yet another reason why we *don't* need a message queue. But to be honest, I think it's only a matter of time till we need our message queue for something, so I'm not super motivated to get rid of Celery.